File: MobileNetV2/src/MobileNetV2.cpp
#include "../inc/MobileNetV2.h"

using namespace std;


MobileNetV2ConvNet::MobileNetV2ConvNet(int num_classes, string model_path)
{
        model = torch::jit::load(model_path);

        // Freeze all parameters in the model
        for (auto& param : model.named_parameters()) {
            param.value().set_requires_grad(false);
        }
}

torch::Tensor MobileNetV2ConvNet::forward(torch::Tensor x)
{
    x = Conv->forward(x);
    x = x.view({x.size(0),-1});
    x = FC->forward(x);
    return x;
}

File: MobileNetV2/MobileNetV2.py
import torch
from torch import nn
import torch.nn.functional as F
import torchvision.models as models



class MobileNet(torch.nn.Module):
    def __init__(self):
        super(MobileNet, self).__init__()
        self.model = models.mobilenet_v2(weights='MobileNet_V2_Weights.DEFAULT')
        
        
        for param in self.model.parameters():
            param.requires_grad = False
            
        
        self.model.classifier[1] = nn.Sequential(
            nn.Linear(in_features=self.model.classifier[1].in_features,out_features=512),
            nn.LeakyReLU(negative_slope=0.02,inplace=True),
            nn.BatchNorm1d(num_features=512),
            nn.Dropout(p=0.4,inplace=True),
            nn.Linear(in_features=512,out_features=2),
            nn.Softmax(dim=1))
        
        # print(self.model)

    def forward(self, x):
        x = self.model(x)
        return x


model = MobileNet()

model.eval()

traced_net = torch.jit.trace(model, torch.randn(1,3,28,28))

torch.jit.save(traced_net,'MobileNet.pt')
